src/java_bytecode/java_bytecode_instrument.cpp

```diff
@@ -0,0 +1,520 @@
+    block.copy_to_operands(instrument_expr(code.op1()));
+    block.copy_to_operands(code);
+
+    code=block;
+  }
+  else if(statement==ID_expression)
+  {
+    INVARIANT(
+      code.has_operands(),
+      "Expression statements should have (1) operands");
+    code_blockt block;
+    block.copy_to_operands(instrument_expr(code.op0()));
+    block.copy_to_operands(code);
+    code=block;
+  }
+  else if(statement==ID_assert)
+  {
+    INVARIANT(
+      code.has_operands(),
+      "Asserts should have (1) operands");
+    // does this correspond to checkcast?
+    if(code.op0().id()==ID_java_instanceof)
+    {
+      code_blockt block;
+      block.copy_to_operands(check_class_cast(
+                               code.op0().op0(),
+                               code.op0().op1(),
+                               code.source_location()));
+      block.copy_to_operands(code);
+      code=block;
+    }
+  }
+  else if(statement==ID_block)
+  {
+    Forall_operands(it, code)
+      instrument_code(to_code(*it));
+  }
+  else if(statement==ID_label)
+  {
+    code_labelt &code_label=to_code_label(code);
+    instrument_code(code_label.code());
+  }
+  else if(statement==ID_ifthenelse)
+  {
+    code_blockt block;
+    code_ifthenelset &code_ifthenelse=to_code_ifthenelse(code);
+    block.copy_to_operands(instrument_expr(code_ifthenelse.cond()));
+    instrument_code(code_ifthenelse.then_case());
+    if(code_ifthenelse.else_case().is_not_nil())
+      instrument_code(code_ifthenelse.else_case());
+    block.copy_to_operands(code);
+    code=block;
+  }
+  else if(statement==ID_switch)
+  {
+    code_blockt block;
+    code_switcht &code_switch=to_code_switch(code);
+    block.copy_to_operands(
+      instrument_expr(code_switch.value()));
+    block.copy_to_operands(code);
+    code=block;
+  }
+  else if(statement==ID_return)
+  {
+    if(code.operands().size()==1)
+    {
+      code_blockt block;
+      block.copy_to_operands(instrument_expr(code.op0()));
+      block.copy_to_operands(code);
+      code=block;
+    }
+  }
+  else if(statement==ID_function_call)
+  {
+    code_blockt block;
+    code_function_callt &code_function_call=to_code_function_call(code);
+    block.copy_to_operands(instrument_expr(code_function_call.lhs()));
+    block.copy_to_operands(instrument_expr(code_function_call.function()));
+
+    for(code_function_callt::argumentst::iterator
+          a_it=code_function_call.arguments().begin();
+        a_it!=code_function_call.arguments().end();
+        a_it++)
+      block.copy_to_operands(instrument_expr(*a_it));
+
+    block.copy_to_operands(code);
+    code=block;
+  }
+}
+
+/// Computes the instrumentation for `expr` in the form of
+/// either assertions or runtime exceptions.
+/// \par parameters: `expr`: the expression for which we compute
+/// strumentation \return: The instrumentation required for `expr`
+codet java_bytecode_instrumentt::instrument_expr(
+  const exprt &expr)
+{
+  if(expr.id()==ID_plus)
```

Comment:
This pattern matching seems likely to accidentally match other expressions in the future... can we identify array accesses more concretely? How about setting a boolean flag on the ID_plus expression during convert_method (ID_java_array_access?) and recognising it here?